coursier.bzl

```diff
@@ -486,70 +487,161 @@ def _coursier_fetch_impl(repository_ctx):
     for a in repository_ctx.attr.excluded_artifacts:
         excluded_artifacts.append(json_parse(a))
 
-    artifact_coordinates = []
-
-    # Set up artifact exclusion, if any. From coursier fetch --help:
-    #
-    # Path to the local exclusion file. Syntax: <org:name>--<org:name>. `--` means minus. Example file content:
-    # com.twitter.penguin:korean-text--com.twitter:util-tunable-internal_2.11
-    # org.apache.commons:commons-math--com.twitter.search:core-query-nodes
-    # Behavior: If root module A excludes module X, but root module B requires X, module X will still be fetched.
-    exclusion_lines = []
-    for a in artifacts:
-        artifact_coordinates.append(utils.artifact_coordinate(a))
-        if "exclusions" in a:
-            for e in a["exclusions"]:
-                exclusion_lines.append(":".join([a["group"], a["artifact"]]) +
-                                       "--" +
-                                       ":".join([e["group"], e["artifact"]]))
-
-    cmd = _generate_coursier_command(repository_ctx)
-    cmd.extend(["fetch"])
-    cmd.extend(artifact_coordinates)
-    cmd.extend(["--artifact-type", ",".join(_COURSIER_PACKAGING_TYPES + ["src"])])
-    cmd.append("--quiet")
-    cmd.append("--no-default")
-    cmd.extend(["--json-output-file", "dep-tree.json"])
-
-    if repository_ctx.attr.fail_on_missing_checksum:
-        cmd.extend(["--checksum", "SHA-1,MD5"])
+    if repository_ctx.attr.maven_install_json:
+        # Read Coursier state from maven_install.json.
+        repository_ctx.symlink(
+            repository_ctx.path(repository_ctx.attr.maven_install_json),
+            repository_ctx.path("imported_maven_install.json")
+        )
+        dep_tree = json_parse(
```

Comment:
By invalid json, do you mean general malformed json or json with corrupted maven_install.json schema?

General malformed json will be handled by json_parse.

Reply:
If I point to a completely unrelated file like a `.py`.
```
ERROR: error loading package '': Encountered error while reading extension file 'defs.bzl': no such package '@mvn//': Traceback (most recent call last):
	File "/home/borkaehw/.cache/bazel/_bazel_borkaehw/66240b6c93cac99d85d9f7c0eb226b35/external/rules_jvm_external/coursier.bzl", line 496
		json_parse(repository_ctx.read(repository_c...")))
	File "/home/borkaehw/.cache/bazel/_bazel_borkaehw/66240b6c93cac99d85d9f7c0eb226b35/external/rules_jvm_external/third_party/bazel_json/lib/json_parser.bzl", line 667, in json_parse
		fail(("parsing JSON failed:\n%s\n\npa...)))
parsing JSON failed:
invalid action: -1  on input '#'

parser:
{"rejected": True, "rejected_reason": "invalid action: -1  on input '#'", "max_depth": 20, "mode_stack": [0], "top": 0, "state": 0, "state_chars": None, "reduction_stack": [[]], "reduction_hooks": {"entry_key": <function _reduce_literal from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "entry_value": <function _reduce_literal from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "object": <function _reduce_object from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "array": <function _reduce_array from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>}, "tokenizer_hooks": {"string": <function _tokenize_string from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "integer": <function _tokenize_int from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "null": <function _tokenize_null from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "true": <function _tokenize_true from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>, "false": <function _tokenize_false from @rules_jvm_external//third_party/bazel_json/lib:json_parser.bzl>}, "hacky_treatment_unknown_chars_as_etc": True, "_DEBUG": False}
```
I feel like the error message is too long to see what's wrong easily.

If it's a corrupted `maven_install.json`. It tosses the entire corrupted file content, which in my case has 20k lines. I can't even see the error message.

It would be nice to just let me know it's the json issue.

Reply:
Thanks for the feedback! Just pushed a commit that improved the JSON parsing error messages.

If it's an invalid maven_install.json:

```
ERROR: An error occurred during the fetch of repository 'maven':
   Failed to parse /usr/local/google/home/jingwen/code/rules_jvm_external/examples/scala_akka/maven_install.json. It is not a valid maven_install.json file. Has this file been modified manually?
```

If it's not a JSON file:

```
$ bazel test //...
INFO: Call stack for the definition of repository 'maven' which is a coursier_fetch (rule definition at /usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/88893d7c3918775765c5e3ae20d931b1/external/rules_jvm_external/coursier.bzl:714:18):
 - /usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/88893d7c3918775765c5e3ae20d931b1/external/rules_jvm_external/defs.bzl:79:9
 - /usr/local/google/home/jingwen/code/rules_jvm_external/examples/scala_akka/WORKSPACE:52:1
DEBUG: /usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/88893d7c3918775765c5e3ae20d931b1/external/rules_jvm_external/third_party/bazel_json/lib/json_parser.bzl:672:13: JSON parsing failed. Could not parse the input load("@io_bazel at: 
...
load("@io_bazel_rules_scala//scala:scala.bzl", "scala_binary", "scala_library", "scala_test_suite")

lib_dependencies = [
    "@maven//:com_chuusai_shapeless_2_12",
    "@maven//:com_github_pureconfig
...
ERROR: An error occurred during the fetch of repository 'maven':
   Failed to parse /usr/local/google/home/jingwen/code/rules_jvm_external/examples/scala_akka/BUILD.bazel. Is this file valid JSON?
```